dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-raft/src/main/java/org/apache/dolphinscheduler/plugin/registry/raft/EphemeralNodeManager.java

```diff
@@ -0,0 +1,405 @@
+
+        @Override
+        public void run() {
+            checkConnection();
+            checkActiveNode();
+        }
+
+        private void checkConnection() {
+            final String host = NetUtils.getHost();
+            if (getActiveMasterServers().keySet().stream().anyMatch(address -> address.split(Constants.COLON)[0].equals(host))) {
+                if (connectionState == null && !connectionListeners.isEmpty()) {
+                    triggerListener(ConnectionState.CONNECTED);
+                } else if (connectionState == ConnectionState.DISCONNECTED) {
+                    triggerListener(ConnectionState.RECONNECTED);
+                } else {
+                    triggerListener(ConnectionState.CONNECTED);
+                }
+                connectionState = ConnectionState.CONNECTED;
+            }
+        }
+
+        private void checkActiveNode() {
+            long expireTime = properties.getConnectionExpireFactor() * properties.getListenerCheckInterval().toMillis();
+            Map<String, Long> activeMasterServers = getActiveMasterServers();
+            for (Map.Entry<String, Long> entry : activeMasterServers.entrySet()) {
+                if ((System.currentTimeMillis() - entry.getValue()) > expireTime) {
+                    final String nodeAddress = entry.getKey();
+                    activeMasterServers.remove(nodeAddress);
+                    updateActiveMaster(activeMasterServers);
+                    addDeadServer(Constants.MASTER_TYPE, nodeAddress);
+                    if (nodeAddress.split(Constants.COLON)[0].equals(NetUtils.getHost())) {
+                        connectionState = ConnectionState.DISCONNECTED;
+                        triggerListener(ConnectionState.DISCONNECTED);
+                        removeNodeData(nodeAddress);
+                    }
+                    log.warn("Master server {} connect to raft cluster timeout, last heartbeat {}, timeout config {} ms",
+                            nodeAddress, convertTimeToString(entry.getValue()), expireTime);
+                }
+            }
+            Map<String, Long> activeWorkerServers = getActiveWorkerServers();
+            for (Map.Entry<String, Long> entry : activeWorkerServers.entrySet()) {
+                if ((System.currentTimeMillis() - entry.getValue()) > expireTime) {
+                    final String nodeAddress = entry.getKey();
+                    activeWorkerServers.remove(nodeAddress);
+                    updateActiveWorker(nodeAddress, activeWorkerServers);
+                    removeWorkerGroup(nodeAddress);
+                    addDeadServer(Constants.WORKER_TYPE, nodeAddress);
+                    removeNodeData(nodeAddress);
+                    log.warn("Worker server {} connect to raft cluster timeout, last heartbeat {}, timeout config {} ms",
+                            nodeAddress, convertTimeToString(entry.getValue()), expireTime);
+                }
+            }
+        }
+
```

Comment:
Here I see `updateActiveMaster` and `updateActiveWorker` method, they will modify whole data under the key `/nodes/master` or `/nodes/worker`, rather than modify the specific record for expired masters found this time.
And still, HeatBeat Task will also do like this.
So there will be multiple threads in different hosts, trying to read and write the same key, which might cause collisions in distributed environment.

My suggestion is : create one key for each master(thus 3 keys for 3 maters), rather than  1 key for 3 masters.